js-library/src/services/ens.ts

```diff
@@ -0,0 +1,176 @@
+    this.publicResolverContract = this.publicResolverContract.connect(signerOrProvider)
+    this.subdomainRegistrarContract = this.subdomainRegistrarContract.connect(signerOrProvider)
+    this.ensRegistryContract = this.ensRegistryContract.connect(signerOrProvider)
+  }
+
+  /**
+   * Returns owner address of the provided username on ENS
+   * @param username ENS username
+   * @returns owner's address
+   */
+  public async getUsernameOwner(username: string): Promise<string> {
+    const usernameHash = this.hashUsername(username)
+
+    const owner = await this.ensRegistryContract.owner(usernameHash)
+
+    return owner
+  }
+
+  /**
+   * Checks whether username is available on ENS
+   * @param username ENS username
+   * @returns True if the username is available
+   */
+  public async isUsernameAvailable(username: string): Promise<boolean> {
+    const owner = await this.getUsernameOwner(username)
+    return owner === NULL_ADDRESS
+  }
+
+  /**
+   * Sets owner of the provided username on ENS
+   * @param username ENS username
+   * @param address Owner address of the username
+   */
+  public async registerUsername(username: string, address: string, publicKey: string): Promise<void> {
+    const ownerAddress = await this.getUsernameOwner(username)
+
+    if (ownerAddress !== NULL_ADDRESS && ownerAddress !== address) {
+      throw new Error(`ENS: Username ${username} is not available`)
+    }
+
+    if (ownerAddress === NULL_ADDRESS) {
+      await waitTransaction(
+        this.subdomainRegistrarContract.register(keccak256(toUtf8Bytes(username)), address),
+      )
+    }
+
+    const usernameHash = this.hashUsername(username)
+
+    await waitTransaction(
+      this.ensRegistryContract.setResolver(usernameHash, this.publicResolverContract.address),
+    )
+
+    await this.setUsernamePublicKey(usernameHash, address, publicKey)
+  }
+
+  /**
+   * Returns public key of registered username
+   * @param username
+   * @returns public key
+   */
+  public async getPublicKey(username: string): Promise<string> {
+    const [publicKeyX, publicKeyY] = await this.publicResolverContract.pubkey(this.hashUsername(username))
+
+    const publicKey = joinPublicKey(publicKeyX, publicKeyY)
+
+    if (isPublicKeyValid(publicKey)) {
+      throw new Error('Public key is not set')
+    }
+
+    return publicKey
+  }
+
+  /**
+   * Fetches all ENS data related to the provided username
+   * @param username ENS username
+   * @returns
+   */
+  public getUserData(username: string): Promise<unknown> {
+    return this.publicResolverContract.getAll(this.hashUsername(username))
+  }
+
+  /**
+   * Sets user's public key to the user's ENS entry
+   * @param username ENS username
+   * @param address Owner of the username
+   * @param publicKey Public key that will be added to ENS
+   */
+  private setUsernamePublicKey(usernameHash: string, address: string, publicKey: string): Promise<void> {
```

Comment:
`usernameHash` has conflict with the JSDoc parameter list.
Btw, I prefer the original idea from the JSDoc for this first parameter

Reply:
The reason why this function expects a hash instead of username is because the hash is already calculated where this function is invoked. So just to avoid calculating hash two times. And because this is a private function, it's not that important.